### PR TITLE
fix(plan): dehydrate the stamped plan.html, not the caller's copy (#810 follow-up)

### PR DESCRIPTION
**Pull request by rsnodgrass**
## Summary

Follow-up to #815. That PR shipped the #810 data-loss fix correctly, but a review caught a latent bug its tests were shaped to miss: **`ox plan save` dehydration silently never runs for a real render**, so large plans still bloat git and every large save leaks an orphaned LFS blob.

**Root cause (in #815):**
- `Save` applies `StampHTMLMeta` (injects `<meta>` tags) and writes the **stamped** bytes to `plan.html`.
- `savePlanArtifacts` then handed `DehydrateHTML` the caller's **unstamped** `html` copy. It uploaded the unstamped bytes, so their OID disagreed with the stamped on-disk file; `guardPointerOverwrite` refused the pointer swap.
- Every real render carries a `<head>`, so this fired every time: dehydration failed, the render stayed plain (the LFS pointer never happened), and the uploaded-but-unreferenced blob was orphaned on each save.
- The #810 guarantee still held — no poisoned pointer, no lost content (the fail-safe worked). Only the feature was dead.

**Fix:** `DehydrateHTML` now reads the bytes it is about to pointerize **straight off disk**, so the upload OID matches the committed file by construction — stamped vs. unstamped can no longer diverge. This also drops the redundant `html` parameter and gates on the actual committed size.

## Test Plan

**The bug fix (`internal/plan`):**
- **`TestSaveThenDehydrate_RealHeadHTML`** drives the true path (`Save` → `DehydrateHTML`) with a >256 KB `<head>`-bearing render and asserts it pointerizes to a stored blob equal to the **stamped** on-disk bytes (explicitly *not* the caller's copy). **Red-first proven**: simulated the mismatch → guard refused → red; restored → green.
- Plus boundary (`== threshold` stays plain), already-a-pointer no-op, nil-client/offline stays-plain, and upload-failure-leaves-plain.

**Deeper coverage the review flagged as missing in #815** (added here so the recovery/guard paths aren't shipping on `ScannedPointers`-only assertions):
- **Reconcile blank+squash** (`TestReconcile_BlanksMissingPlanPointerAndSquashes`) — a 404 plan blob is actually blanked to empty and squashed out of the unpushed pack (bare remote + fake LFS server, via a new client-injectable reconcile seam).
- **Transient-error guard** (`TestReconcile_TransientErrorNeverBlanks`) — 401/503 aborts, never zeroes a live pointer, for **both** the plan and session trees.
- **Composition-root e2e** (`TestSavePlanArtifacts_LargeRenderStaysPlainWhenLFSUnreachable`) — a large render through the real `savePlanArtifacts` stays plain (retrievable, pushable, no poisoned pointer) when no LFS client is reachable.
- **Doctor warn/fix wiring** (`TestEvaluatePlanPointers_WarnThenFix`) — no-`--fix` warns and never reconciles; `--fix` invokes the reconcile and reports its result.
- **`UploadBlob` already-present** yields a usable proof; **`TestNoAssertUploadedOfFreshFileRef`** is a source-scan enforcement test that fails if `AssertUploaded(NewFileRef(…))` (minting proof for un-uploaded content) ever reappears — the backstop the compile-time type can't provide.

`go build ./...` clean; `internal/lfs`, `internal/plan`, and the cmd/ox `Plan`/doctor suites green; `go vet` clean on all touched packages.

Follows #815 · closes the dehydration gap in #810.

---
<details>
<summary>Checklist (expand if needed)</summary>

- [x] Tests added (red-first regression + fault/boundary cases)
- [ ] CHANGELOG — N/A: fixes an unshipped-behavior gap in an experimental, off-by-default surface
- [x] Breaking change — none: internal signature only (`DehydrateHTML` dropped its `html` param)
- [x] `/security-review` — N/A: no auth/mcp/raw_writer/prepush/upgrade/go.sum; LFS stays pure-Go

</details>

Co-Authored-By: [SageOx](https://github.com/SageOx)

SageOx-Session: https://sageox.ai/c/ses_01a03520-1293-7664-ba71-f255ef40acee

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved large-file handling so uploaded content matches the stamped plan file exactly.
  * Preserved plan files when reading or uploading fails.
  * Avoided unnecessary processing for missing, unreadable, already-pointerized, offline, or undersized files.
  * Improved plan-pointer checks and repairs, with safer handling of missing or temporarily unavailable blobs.
  * Prevented invalid pointers when uploaded content is unavailable.

* **Documentation**
  * Added guidance for conversation workflows and pull request header generation.

### CODE REVIEW

File: internal/plan/dehydrate.go [hunk 1]
```diff
@@ -2,6 +2,7 @@ package plan
 
 import (
 	"fmt"
+	"os"
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply internal/plan/dehydrate.go:7]: Fixed. `DehydrateHTML` now reads the file first and only treats `fs.ErrNotExist` as a no-op; every other read error is wrapped and returned (`b4882660`).

File: internal/plan/dehydrate.go [hunk 2]
```diff
@@ -12,36 +13,56 @@ import (
...
+//   - upload succeeds: overwrites plan.html with a pointer to the now-present blob.
+func DehydrateHTML(dir string, client *lfs.Client) (bool, error) {
+	htmlPath := filepath.Join(dir, planHTMLFile)
+
+	info, err := os.Stat(htmlPath)
+	if err != nil {
+		return false, nil // no render on disk (or unreadable) — nothing to dehydrate
+	}
+	if info.Size() <= htmlLFSThreshold {
+		return false, nil // small enough to keep plain
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply internal/plan/dehydrate.go]: Fixed. Removed the separate `os.Stat`; the size gate is now on `len(content)` of the exact bytes read and uploaded, so no stat-vs-read snapshot can diverge under a concurrent `Save` (`b4882660`).

File: cmd/ox/doctor_plan_pointers_test.go [hunk 3]
```diff
@@ -97,3 +97,79 @@ func TestCollectPlanHTMLPointers_SkipsPlainAndNonPlans(t *testing.T) {
...
+	// fix path: reconcile is invoked and its result is reported as a pass.
+	called := false
+	fix := evaluatePlanPointers(client, pointers, true, func() (*lfs.ReconcileResult, error) {
+		called = true
+		require.NoError(t, os.WriteFile(htmlPath, []byte{}, 0o644)) // simulate the blank
+		return &lfs.ReconcileResult{Replaced: 1}, nil
+	})
+	assert.True(t, called, "the --fix path must invoke the reconcile")
+	assert.False(t, fix.warning, "a successful reconcile is a passed result")
+	assert.Contains(t, fix.message, "reconciled")
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply cmd/ox/doctor_plan_pointers_test.go:175]: Fixed. Added a reconcile-failure sub-case asserting `evaluatePlanPointers` returns a warning containing `"reconcile failed"` with the missing count preserved (`b4882660`).

File: cmd/ox/plan_capture_test.go [hunk 4]
```diff
@@ -380,3 +382,40 @@ func TestPlanSessionLink_DisabledAttributionExpectsNoLink(t *testing.T) {
...
+	if isPointer {
+		t.Fatalf("large render was committed as an LFS pointer with no reachable store — a poisoned pointer (GH #810)")
+	}
+	got, err := os.ReadFile(path)
+	if err != nil {
+		t.Fatalf("read plan.html: %v", err)
+	}
+	if !bytes.Contains(got, []byte("PRESERVE-ME")) {
+		t.Fatalf("plan.html does not contain the render content — not retrievable")
+	}
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply cmd/ox/plan_capture_test.go:424]: This harness configures no git ledger, so there is no commit to assert against with `git show`. `DehydrateHTML` rewrites `plan.html` BEFORE `commitPlanToLedger` stages the working tree, so the on-disk plain file is exactly what a commit would carry here. Strengthened the payload check instead (sibling thread).

File: cmd/ox/plan_capture_test.go [hunk 5]
```diff
@@ -380,3 +382,40 @@ func TestPlanSessionLink_DisabledAttributionExpectsNoLink(t *testing.T) {
...
+	if isPointer {
+		t.Fatalf("large render was committed as an LFS pointer with no reachable store — a poisoned pointer (GH #810)")
+	}
+	got, err := os.ReadFile(path)
+	if err != nil {
+		t.Fatalf("read plan.html: %v", err)
+	}
+	if !bytes.Contains(got, []byte("PRESERVE-ME")) {
+		t.Fatalf("plan.html does not contain the render content — not retrievable")
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply cmd/ox/plan_capture_test.go]: Fixed. Replaced `bytes.Contains` with `bytes.Count(got, marker) == 30000`, proving the full render survived rather than a single stray marker (`b4882660`).

File: internal/lfs/uploaded_test.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+			}
+			return nil
+		}
+		if !strings.HasSuffix(path, ".go") || strings.HasSuffix(path, "_test.go") {
+			return nil
+		}
+		data, rerr := os.ReadFile(path)
+		if rerr != nil {
+			return nil
+		}
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply internal/lfs/uploaded_test.go:89]: Fixed. The `WalkDir` callback now returns wrapped errors on traversal and read failures, so an unreadable production file fails the scan via `require.NoError` instead of passing vacuously (`b4882660`).
